inject/src/main/java/io/micronaut/context/env/PropertySourcePropertyResolver.java

```diff
@@ -84,14 +84,22 @@ public class PropertySourcePropertyResolver implements PropertyResolver, AutoClo
     protected Logger log;
 
     private final Map<String, Boolean> containsCache = new ConcurrentHashMap<>(20);
-    private final Map<String, Object> resolvedValueCache = new ConcurrentHashMap<>(20);
+    /**
+     * Cache for values <i>before</i> conversion. This avoids recomputing placeholders, which keeps
+     * random values (e.g. {@code ${random.port}} stable).
+     */
+    private final Map<String, Object> placeholderResolutionCache = new ConcurrentHashMap<>(20);
+    /**
+     * Cache for values <i>after</i> conversion.
+     */
+    private final Map<ConversionCacheKey, Object> resolvedValueCache = new ConcurrentHashMap<>(20);
     private final EnvironmentProperties environmentProperties = EnvironmentProperties.fork(CURRENT_ENV);
 
     /**
      * Creates a new, initially empty, {@link PropertySourcePropertyResolver} for the given {@link ConversionService}.
      *
      * @param conversionService The {@link ConversionService}
-     * @param logEnabled logEnabled flag to enable or disable logger
+     * @param logEnabled        logEnabled flag to enable or disable logger
```

Comment:
That's my mistake `logEnabled` word twice. Could you fix it?

Reply:
```suggestion
     * @param logEnabled        flag to enable or disable logger
```